### PR TITLE
Fix payload and defaults

### LINKED ISSUE #62450: [BUG] 3005 pillar.get, vault.read_secret, vault.list_secrets changes are slightly problematic
**Description**

Severial modules were updated in 3005 to change their default behavior of the... `default` parameter: https://github.com/saltstack/salt/pull/61089

There are several issues with that change:
1. Change is not documented. Nothing states that if the **default** is **None** it will lead to the error raise.
2. Changing the defaults should be done very carefully and usually with some kind of deprecation policy. People could rely on defaults. Like we used `default=None` as a secret default for the vault! It just makes sense to us.

**Expected behavior**
I do not fully understand the issue that was attempted to be fixed here so I can't really suggest the change here.

### PR DESCRIPTION
**Pull request by twangboy**
### What does this PR do?
Adds a new Constant `NOT_SET` to allow `None` to be passed as a valid option to bypass throwing an error. We can't set default to be an object because msgpack can't serialize it and causes `sys.argparse` to fail when returning data to the master.

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/62450

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

### CODE REVIEW

**Review by s0undt3ch — DISMISSED** [review]

**Review by Ch3LL — CHANGES_REQUESTED** [review]
changelog? and does there need to be test coverage?

**Comment by twangboy** [comment]
There was a test written in the original PR. I think it covers the changes here:
https://github.com/saltstack/salt/pull/61089

**Comment by Ch3LL** [comment]
It doesn't look like those tests are covering the issue created in https://github.com/saltstack/salt/issues/62450 . It should be easy to add a test that ensures none is returned when default none is passed. 

`salt-call --local pillar.get doesnotexist default=None`

**Comment by frebib** [comment]
I tend to agree with Wayne's comment (https://github.com/saltstack/salt/issues/62450#issuecomment-1212262538) suggesting to use a static constant instead of magic behaviour based on internal python values to avoid any chance of conflict or ambiguity. Arguably using `None` in the first place is the same idea as using `...` but just with a more commonly used constant

**Comment by s0undt3ch** [comment]
> I tend to agree with Wayne's comment ([#62450 (comment)](https://github.com/saltstack/salt/issues/62450#issuecomment-1212262538)) suggesting to use a static constant instead of magic behaviour based on internal python values to avoid any chance of conflict or ambiguity. Arguably using `None` in the first place is the same idea as using `...` but just with a more commonly used constant

A static constant would involve more code to support it(more imports in the code base).

`...` is not magical, [it's a built-in constant](https://docs.python.org/3/library/constants.html#Ellipsis), and in this context, it serves the exact same purpose as `None` without the side-effects that `None` caused, and with no additional imports in the code.

**Review by s0undt3ch — DISMISSED** [review]

**Comment by s0undt3ch** [comment]
Would this cause less potential confusion if we used `False` instead of `...`?

**Comment by s0undt3ch** [comment]
After some internal discussion, I retracted my proposal and we implemented a "constant".

**Review by Ch3LL — APPROVED** [review]

File: salt/defaults/__init__.py [hunk 1]
```diff
@@ -5,5 +5,58 @@
...
+
+    def __hash__(self):
+        return hash((self.name, self.value))
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return False
+        if self.name != other.name:
+            return False
+        return self.value == other.value
```
Comment (waynew) [inline salt/defaults/__init__.py:35]: **suggestion (non-blocking)** could be `return self.name == other.name and self.value == other.value`, but that's not an essential change

**Review by waynew — APPROVED** [review]
One thought, but it's not a blocker at all. This looks good - I didn't run the tests locally, so trusting that happened correctly on your end :slightly_smiling_face:
